### DISCUSSION TITLE
User Control events not firing when rendered in a DataTemplate

### DISCUSSION THREAD

**Discussion by mathtone** [AvaloniaUI/Avalonia#15884] [Q&A] [answered]
I have simple TabControl, with an ItemsSource bound to an ObservableCollection in the VM, everything seems to work for the first control, and as additional items are added new tabs are created and bound to their ViewModel(s) but it seems as though no events are fired on any but the first one to be created, I've tried "Loaded" and "Initialized" and "AttachedTo(Whatever)Tree" but it really seems as though none of these are firing, even if I override the method in the view's code-behind or manually attach an event handler I'm pretty stumped....  any advice:

MainView:

<UserControl xmlns="https://github.com/avaloniaui"
			 xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
			 xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
			 xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
			 xmlns:vm="clr-namespace:AvaloniaStudy.ViewModels"
			 xmlns:local="clr-namespace:AvaloniaStudy.Views"
			 xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
			 xmlns:ia="clr-namespace:Avalonia.Xaml.Interactions.Core;assembly=Avalonia.Xaml.Interactions"
			 mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
			 x:Class="AvaloniaStudy.Views.MainView"
			 x:DataType="vm:MainViewModel">
	
	<DockPanel>

		<StackPanel Orientation="Horizontal" DockPanel.Dock="Top">
			<TextBlock Text="{Binding Time}" HorizontalAlignment="Center" VerticalAlignment="Center" />
			<Button Content="Add" Command="{Binding ActivateTestCommand}" />
		</StackPanel>

		<TabControl ItemsSource="{Binding TestViewModels}">
			<TabControl.ItemTemplate>
				<DataTemplate >
					<TextBlock Text="BEEP" Foreground="Orchid"/>
				</DataTemplate>
			</TabControl.ItemTemplate>
			<TabControl.ContentTemplate>
				<DataTemplate DataType="vm:TestViewModel">
					<local:TestView/>
				</DataTemplate>
			</TabControl.ContentTemplate>
		</TabControl>
	</DockPanel>
</UserControl>

TestView:
<UserControl xmlns="https://github.com/avaloniaui"
			 xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
			 xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
			 xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
			 xmlns:vm="clr-namespace:AvaloniaStudy.ViewModels"
			 xmlns:i="clr-namespace:Avalonia.Xaml.Interactivity;assembly=Avalonia.Xaml.Interactivity"
			 xmlns:ia="clr-namespace:Avalonia.Xaml.Interactions.Core;assembly=Avalonia.Xaml.Interactions"
			 mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
			 x:Class="AvaloniaStudy.Views.TestView"
			
			 x:DataType="vm:TestViewModel">
	<i:Interaction.Behaviors>
		<ia:EventTriggerBehavior EventName="AttachedToVisualTree">
			<ia:InvokeCommandAction Command="{Binding OnLoadCommand}" />
		</ia:EventTriggerBehavior>
	</i:Interaction.Behaviors>
	<StackPanel>
		<TextBlock Text="{Binding Name}"/>
		<TextBlock Text="{Binding Title}"/>
	</StackPanel>
</UserControl>

**Comment by timunie** [comment, marked as answer]
What happens if you change the tab? Content is only set once for a single item and not for all items. If you activate another tab, Id think the loaded event may be raised again. If not the View was recycled and in that case you can listen to Data context changed event

**Reply by mathtone** [reply]
Switching back & forth doesn't make a difference but the DataContextChanged does seem to fire every time the tabs are switched, I might be able to build a workaround based on that.  Even then it doesn't *seem* to do it the first time, I have to switch away from the tab and back again....  I might be misreading something.  Still pretty new w/ Avalonia & getting used to it.

**Reply by timunie** [reply]
I don't even understand why you need that event. Normally, it should without these events. Can you explain what the event should do?

**Reply by mathtone** [reply]
The Viewmodel needs to get initialized and call some services to retrieve various lists, etc.  Normally in WPF this is how I'd have done that,  when the control gets "Loaded" it calls a command in the VM to initialize itself.  Is there another recommended approach?

**Reply by timunie** [reply]
I use to bin to selectedItem and if that changes, the ViewModel should call init. That way I've decopuled it from the view as good as possible. That's probably philosophy anyways.
